### PR TITLE
Multicard support for rocm platform

### PR DESCRIPTION
**Pull request by jichangjichang**

### CODE REVIEW

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 1]
```diff
@@ -478,7 +478,7 @@ bool MultiDevSSAGraphBuilder::IsSparseGradient(const std::string &og) const {
 
 void MultiDevSSAGraphBuilder::SetCommunicationContext(
     OpHandleBase *op_handle, const platform::Place &p) const {
-#if (defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)) && !defined(_WIN32)
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
   if (nccl_ctxs_ == nullptr) {
     op_handle->SetDeviceContext(p,
```
Comment (sabreshao) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc:483]: This line is not needed in HIP?

File: paddle/fluid/framework/parallel_executor.cc [hunk 2]
```diff
@@ -351,6 +351,8 @@ ParallelExecutor::~ParallelExecutor() {
   for (auto &p : member_->places_) {
     platform::DeviceContextPool::Instance().Get(p)->Wait();
   }
+  member_->nccl_ctxs_.reset(NULL);
```
Comment (sabreshao) [inline paddle/fluid/framework/parallel_executor.cc]: These two is to fix Paddle bug? Please add comment to note this.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 1 above)
Comment (jichangjichang) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc:483]: This change is for single stream

File: paddle/fluid/framework/parallel_executor.cc (hunk 2 above)
Comment (jichangjichang) [reply paddle/fluid/framework/parallel_executor.cc]: Fix memory leakage. I will add later.

**Pull request merged**